### PR TITLE
feat(dbt): add `get_partition_mapping` to `DagsterDbtTranslator`

### PR DESCRIPTION
**Pull request by rexledesma**
## Summary & Motivation
Resolves https://github.com/dagster-io/dagster/issues/13053.

Allow `PartitionMapping` to be defined for dbt root dependencies (e.g. dbt sources, dbt models with no dependencies). In a follow-up PR, we will add a way to define self-dependencies on dbt models, so that `PartitionMapping`'s can be defined on a model's previous partitions rather than the model's inputs.

## How I Tested These Changes
add a test

### CODE REVIEW

**Comment by rexledesma** [comment]
Current dependencies on/for this PR:
* master
  * **PR #17480** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17480" target="_blank">Graphite</a> 
    * **PR #17488** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/17488" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/17488?utm_source=stack-comment).

**Review by sryza — COMMENTED** [review]
If I understand correctly, this assumes a single partition mapping per dbt model.  However, different dependencies of the same model can have different partition mappings.  E.g. a common case is that model might depend on itself as well as a different upstream model, and the partition mapping for itself might have an offset, but the partition mapping for the other model might not have that offset.

An option to consider here: underneath the `has_self_dependency` `+meta` tag, allow specifying the start and end offset. We currently only allow self dependencies when there'sa time window partition mapping with a start and end offset.

**Comment by rexledesma** [comment]
@sryza Got it. Here's the current state of things:

1. One `@dbt_assets` corresponds to a single `PartitionsDefinition`.
2. Each entry in `internal_asset_deps` associated in `@dbt_assets` can have a separate `PartitionMapping`.
3. Adding a self-dependency adds another entry in `internal_asset_deps`, with the constraint that a `TimeWindowPartitionMapping` is enforced for its `PartitionMapping`.

The current solution addresses (3), but not (2). My ideal solution would:
- Address both (2) and (3)
- Does not resort to serializing `PartitionMapping` in the dbt `+meta`. I don't know in the future if `TimeWindowPartitionMapping` will be subclassed in the future, or other user-defined partition mappings will be valid here. We should prefer to encode this in code.

Some thoughts:
- To address (2), `DagsterDbtTranslator.get_partition_mapping` should take two parameters (`dbt_resource_props`, `dbt_parent_resource_props`). This way, each entry in `internal_asset_deps` could receive its own custom `PartitionMapping`. In the case that this is a self-dependency, `dbt_resource_props` and `dbt_parent_resource_props` would be the same.
- How common is configuring (2), which is the general case for (3)? I imagine this would also be useful for configuring `PartitionMapping` to assets defined for dbt sources. But internal dependencies within the `@dbt_assets` would probably have the same `PartitionMapping`.

**Comment by sryza** [comment]
> Does not resort to serializing PartitionMapping in the dbt +meta. I don't know in the future if TimeWindowPartitionMapping will be subclassed in the future, or other user-defined partition mappings will be valid here. We should prefer to encode this in code.

I'm of two minds on this. It's not legal to have a self-dependency without specifying an offsetted partition mapping for it, so, in one way of looking at it, specifying just `has_self_dependency` on its own is incomplete.

> To address (2), DagsterDbtTranslator.get_partition_mapping should take two parameters (dbt_resource_props, dbt_parent_resource_props). This way, each entry in internal_asset_deps could receive its own custom PartitionMapping. In the case that this is a self-dependency, dbt_resource_props and dbt_parent_resource_props would be the same.

This makes sense to me.  Whatever we do for self dependencies, I think we'll need to have this to support more general partition mapping for dependencies.

> How common is configuring (2), which is the general case for (3)? I imagine this would also be useful for configuring PartitionMapping to assets defined for dbt sources. But internal dependencies within the @dbt_assets would probably have the same PartitionMapping.

I think it's very possible that an internal dependency within the @dbt_assets has a different partition mapping than other internal dependencies. For example, if they're all daily partitioned, but one depends on a rolling window of previous days in an upstream internal asset.

File: python_modules/libraries/dagster-dbt/dagster_dbt/dagster_dbt_translator.py [hunk 1]
```diff
@@ -115,6 +116,35 @@ def get_asset_key(cls, dbt_resource_props: Mapping[str, Any]) -> AssetKey:
...
+        This method can be overridden to provide a custom partition mapping for a dbt dependency.
+
+        Args:
+            dbt_resource_props (Mapping[str, Any]):
+                A dictionary representing the dbt child resource.
+            dbt_parent_resource_props (Mapping[str, Any]):
+                A dictionary representing the dbt parent resource, in relationship to the child.
+
+        Returns:
+            Optional[PartitionMapping]: The Dagster partition mapping for the dbt resource.
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/dagster_dbt_translator.py:144]: I think it's worth clarifying what returning ` None` means. I assume it means we'll use the default partition mapping?

**Review by sryza — APPROVED** [review]
Nice

**Pull request merged**